src/category_theory/sites/pretopology.lean

```diff
@@ -0,0 +1,274 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+
+import category_theory.sites.grothendieck
+
+/-!
+# Grothendieck pretopologies
+
+Definition and lemmas about Grothendieck pretopologies.
+A Grothendieck pretopology for a category `C` is a set of families of morphisms with fixed codomain,
+satisfying certain closure conditions.
+
+We show that a pretopology generates a genuine Grothendieck topology, and every topology has
+a maximal pretopology which generates it.
+
+## Todo
+
+Define sheaves on a pretopology, and show they are the same as the sheaves for the topology
+generated by the pretopology.
+
+## Tags
+
+coverage, pretopology, site
+
+## References
+
+* [https://ncatlab.org/nlab/show/Grothendieck+pretopology][nlab]
+* [S. MacLane, I. Moerdijk, *Sheaves in Geometry and Logic*][MM92]
+* [https://stacks.math.columbia.edu/tag/00VG][Stacks]
+-/
+
+universes v u
+noncomputable theory
+
+namespace category_theory
+
+open category_theory category limits arrows_with_codomain
+
+variables {C : Type u} [category.{v} C] [has_pullbacks C]
+
+/--
+Pullback a set of arrows with given codomain along a fixed map, by taking the pullback in the
+category.
+This is not the same as the arrow set of `sieve.pullback`, but there is a relation between them
+in `pullback_arrows_comm`.
+-/
+def pullback_arrows {X Y : C} (f : Y ⟶ X) (S : arrows_with_codomain X) :
+  arrows_with_codomain Y :=
+λ Z g, ∃ Z' (h : Z' ⟶ X), S h ∧ ∃ (H : Z = pullback h f),
+  eq_to_hom H ≫ pullback.snd = g
+
+lemma pullback_arrows_comm {X Y : C} (f : Y ⟶ X)
+  (R : arrows_with_codomain X) :
+  sieve.generate (pullback_arrows f R) = (sieve.generate R).pullback f :=
+begin
+  ext Z g,
+  split,
+  { rintro ⟨W, k, l, ⟨T, g, hg, rfl, rfl⟩, rfl⟩,
+    refine ⟨_, k ≫ pullback.fst, g, hg, _⟩,
+    rw [assoc, pullback.condition, eq_to_hom_refl, id_comp, assoc] },
+  { rintro ⟨W, k, h, hh, comm⟩,
+    exact ⟨_, pullback.lift _ _ comm, _, ⟨_, h, hh, rfl, rfl⟩, by simp⟩ },
+end
+
+lemma pullback_singleton {X Y Z : C} (f : Y ⟶ X) (g : Z ⟶ X) :
+ ∃ (W : C) (k : W ⟶ Y), pullback_arrows f (singleton_arrow g) = singleton_arrow k :=
+begin
+  refine ⟨pullback (eq_to_hom (eq.refl _) ≫ g) f, pullback.snd, _⟩,
+  ext W k,
+  split,
+  { rintro ⟨W, k, ⟨rfl, rfl⟩, rfl, rfl⟩,
+    rw [eq_to_hom_refl (pullback (eq_to_hom (eq.refl W) ≫ g) f) (eq.refl _), id_comp],
+    apply singleton_arrow_self },
+  { rintro ⟨rfl, rfl⟩,
+    exact ⟨_, _, by simp, _, rfl⟩ }
+end
+
+variables (C)
+
+/--
+A (Grothendieck) pretopology on `C` is a collection of morphisms with fixed target for each object,
```

Comment:
I think this comment could be reworded to make it a bit clearer. E.g. the first sentence could say

```
A (Grothendieck) pretopology on `C` consists of a collection of families of morphisms with a fixed target `X` for every object `X` in `C`, called "coverings" of `X`, which satisfies the following three axioms:
```

Reply:
Looks good to me, thanks.

Reply:
I agree that it would be good to reformulate as Adam suggests.